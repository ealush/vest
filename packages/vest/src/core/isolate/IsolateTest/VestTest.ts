import { invariant, isPromise, optionalFunctionValue } from 'vest-utils';

import type { TIsolateTest } from 'IsolateTest';
import {
  TestAction,
  TestStateMachineAction,
  TestStatus,
  createTestStateMachine,
} from 'IsolateTestStateMachine';
import { TestSeverity } from 'Severity';
import { TFieldName, TGroupName } from 'SuiteResultTypes';

const TestStateMachine = createTestStateMachine();

export class VestTest {
  // Read

  static getData<
    F extends TFieldName = TFieldName,
    G extends TGroupName = TGroupName
  >(test: TIsolateTest<F, G>) {
    invariant(test.data);
    return test.data;
  }

  static warns(test: TIsolateTest): boolean {
    return VestTest.getData(test).severity === TestSeverity.Warning;
  }

  static isPending(test: TIsolateTest): boolean {
    return VestTest.statusEquals(test, TestStatus.PENDING);
  }

  static isOmitted(test: TIsolateTest): boolean {
    return VestTest.statusEquals(test, TestStatus.OMITTED);
  }

  static isUntested(test: TIsolateTest): boolean {
    return VestTest.statusEquals(test, TestStatus.UNTESTED);
  }

  static isFailing(test: TIsolateTest): boolean {
    return VestTest.statusEquals(test, TestStatus.FAILED);
  }

  static isCanceled(test: TIsolateTest): boolean {
    return VestTest.statusEquals(test, TestStatus.CANCELED);
  }

  static isSkipped(test: TIsolateTest): boolean {
    return VestTest.statusEquals(test, TestStatus.SKIPPED);
  }

  static isPassing(test: TIsolateTest): boolean {
    return VestTest.statusEquals(test, TestStatus.PASSING);
  }

  static isWarning(test: TIsolateTest): boolean {
    return VestTest.statusEquals(test, TestStatus.WARNING);
  }

  static hasFailures(test: TIsolateTest): boolean {
    return VestTest.isFailing(test) || VestTest.isWarning(test);
  }

  static isNonActionable(test: TIsolateTest): boolean {
    return (
      VestTest.isSkipped(test) ||
      VestTest.isOmitted(test) ||
      VestTest.isCanceled(test)
    );
  }

  static isTested(test: TIsolateTest): boolean {
    return VestTest.hasFailures(test) || VestTest.isPassing(test);
  }

  static awaitsResolution(test: TIsolateTest): boolean {
    // Is the test in a state where it can still be run, or complete running
    // and its final status is indeterminate?
    return (
      VestTest.isSkipped(test) ||
      VestTest.isUntested(test) ||
      VestTest.isPending(test)
    );
  }

  static isAsyncTest(test: TIsolateTest): boolean {
    return isPromise(VestTest.getData(test).asyncTest);
  }

  static statusEquals(test: TIsolateTest, status: TestStatus): boolean {
    return VestTest.getData(test).status === status;
  }

  // Mutate

  static setPending(test: TIsolateTest) {
    VestTest.setStatus(test, TestStatus.PENDING);
  }

  static fail(test: TIsolateTest): void {
    VestTest.setStatus(
      test,
      VestTest.warns(test) ? TestStatus.WARNING : TestStatus.FAILED
    );
  }

  static pass(test: TIsolateTest): void {
    VestTest.setStatus(test, TestStatus.PASSING);
  }

  static warn(test: TIsolateTest): void {
    VestTest.setData(test, current => ({
      ...current,
      severity: TestSeverity.Warning,
    }));
  }

  static setData(
    test: TIsolateTest,
    setter:
      | ((current: TIsolateTest['data']) => TIsolateTest['data'])
      | TIsolateTest['data']
  ): void {
    test.data = optionalFunctionValue(setter, VestTest.getData(test));
  }

  static skip(test: TIsolateTest, force?: boolean): void {
    // Without this force flag, the test will be marked as skipped even if it is pending.
    // This means that it will not be counted in "allIncomplete" and its done callbacks
    // will not be called, or will be called prematurely.
    // What this mostly say is that when we have a pending test for one field, and we then
    // start typing in a different field - the pending test will be canceled, which
    // is usually an unwanted behavior.
    // The only scenario in which we DO want to cancel the async test regardless
    // is when we specifically skip a test with `skipWhen`, which is handled by the
    // "force" boolean flag.
    // I am not a fan of this flag, but it gets the job done.
    VestTest.setStatus(test, TestStatus.SKIPPED, force);
  }

  static cancel(test: TIsolateTest): void {
    VestTest.setStatus(test, TestStatus.CANCELED);
    VestTest.getData(test).abortController.abort(TestStatus.CANCELED);
  }

  static omit(test: TIsolateTest): void {
    VestTest.setStatus(test, TestStatus.OMITTED);
  }

  static reset(test: TIsolateTest): void {
    VestTest.setStatus(test, TestAction.RESET);
  }

  static setStatus(
    test: TIsolateTest,
    status: TestStateMachineAction,
    payload?: any
  ): void {
    VestTest.setData(test, current => ({
      ...current,
      status: TestStateMachine.staticTransition(
        current.status,
        status,
        payload
      ),
    }));
  }
}
