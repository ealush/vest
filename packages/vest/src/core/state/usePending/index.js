import state from '..';

export default state.registerHandler(() => ({
  pending: [],
  lagging: [],
}));
