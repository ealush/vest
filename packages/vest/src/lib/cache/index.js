/**
 * Creates a cache function
 * @param {number} [maxSize] Max cache size
 * @return {Function} cache function
 */
const createCache = (maxSize = 10) => {
  const cacheStorage = [];

  /**
   * @param {Any[]} deps  dependency array.
   * @param {Function}    cache action function.
   */
  const cache = (deps, cacheAction) => {
    const cacheHit = cache.get(deps);

    if (cacheHit !== null) {
      return cacheHit[1];
    }

    const result = cacheAction();

    cacheStorage.unshift([[...deps], result]);

    if (cacheStorage.length > maxSize) {
      cacheStorage.length = maxSize;
    }

    return result;
  };

  /**
   * Retrieves an item from the cache.
   * @param {deps} deps Dependency array
   */
  cache.get = deps =>
    cacheStorage[
      cacheStorage.findIndex(
        ([cachedDeps]) =>
          deps.length === cachedDeps.length &&
          deps.every((dep, i) => dep === cachedDeps[i])
      )
    ] ?? null;

  return cache;
};

export default createCache;
