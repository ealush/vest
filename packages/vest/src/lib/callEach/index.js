export default function callEach(arr) {
  return arr.forEach(fn => fn());
}
