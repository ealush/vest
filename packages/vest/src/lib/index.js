export { default as globalObject } from "./globalObject";
export { default as runWithContext } from "./runWithContext";
export { default as singleton } from "./singleton";
export { default as throwError } from "./throwError";
