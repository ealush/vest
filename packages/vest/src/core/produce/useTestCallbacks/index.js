import state from '../../state';

export default state.registerHandler(() => ({
  fieldCallbacks: [],
  doneCallbacks: [],
}));
