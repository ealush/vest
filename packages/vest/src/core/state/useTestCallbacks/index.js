import state from '..';

export default state.registerHandler(() => ({
  fieldCallbacks: [],
  doneCallbacks: [],
}));
