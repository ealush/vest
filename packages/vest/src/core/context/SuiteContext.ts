import { createCascade } from 'context';
import { assign, TinyState, tinyState, cache, CacheApi } from 'vest-utils';
import { DynamicValue } from 'vest-utils/src/utilityTypes';

import { IsolateTest } from 'IsolateTest';
import { Modes } from 'Modes';

export const SuiteContext = createCascade<CTXType>((ctxRef, parentContext) => {
  if (parentContext) {
    return null;
  }

  return assign(
    {
      inclusion: {},
      mode: tinyState.createTinyState<Modes>(Modes.EAGER),
      suiteParams: [],
      testMemoCache,
    },
    ctxRef
  );
});

type CTXType = {
  inclusion: Record<string, DynamicValue<boolean>>;
  mode: TinyState<Modes>;
  suiteParams: any[];
  testMemoCache: CacheApi<IsolateTest>;
  currentTest?: IsolateTest;
  groupName?: string;
  skipped?: boolean;
  omitted?: boolean;
};

export function useCurrentTest(msg?: string) {
  return SuiteContext.useX(msg).currentTest;
}

export function useGroupName() {
  return SuiteContext.useX().groupName;
}

export function useInclusion() {
  return SuiteContext.useX().inclusion;
}

export function useMode() {
  return SuiteContext.useX().mode();
}

export function useSkipped() {
  return SuiteContext.useX().skipped ?? false;
}

export function useOmitted() {
  return SuiteContext.useX().omitted ?? false;
}

const testMemoCache = cache<IsolateTest>(10);

export function useTestMemoCache() {
  return SuiteContext.useX().testMemoCache;
}

export function useSuiteParams() {
  return SuiteContext.useX().suiteParams;
}
