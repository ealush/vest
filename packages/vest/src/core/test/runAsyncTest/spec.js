import _ from 'lodash';
import runCreateRef from '../../../../testUtils/runCreateRef';
import context from '../../context';
import useTestCallbacks from '../../produce/useTestCallbacks';
import VestTest from '../lib/VestTest';
import { setPending } from '../lib/pending';
import usePending from '../lib/pending/usePending';
import runAsyncTest from '.';

const STATEMENT = 'some statement string';

const CASE_PASSING = 'passing';
const CASE_FAILING = 'failing';

let stateRef;

it.ctx = (str, cb) => it(str, () => context.run({ stateRef }, cb));

describe.each([CASE_PASSING /*, CASE_FAILING*/])(
  'runAsyncTest: %s',
  testCase => {
    let testObject, fieldName;

    const runRunAsyncTest = (...args) =>
      context.run(
        {
          stateRef,
        },
        () => runAsyncTest(...args)
      );

    beforeEach(() => {
      fieldName = 'field_1';

      stateRef = runCreateRef();
      context.run({ stateRef }, () => {
        const [, setTestCallbacks] = useTestCallbacks();
        setTestCallbacks(state => {
          state.fieldCallbacks = {
            ...state.fieldCallbacks,
            [fieldName]: state.fieldCallbacks[fieldName] || [],
          };

          return state;
        });
      });
      testObject = new VestTest({
        fieldName,
        statement: STATEMENT,
        testFn: () => null,
      });
      testObject.asyncTest =
        testCase === CASE_PASSING ? Promise.resolve() : Promise.reject();
      context.run({ stateRef }, () => {
        setPending(testObject);
      });
    });

    describe('State updates', () => {
      it.ctx('Initial state matches snapshot (sanity)', () => {
        const [pendingState] = usePending();
        expect(pendingState.pending).toContain(testObject);
        expect(stateRef.current()).toMatchSnapshot();
        runRunAsyncTest(testObject);
      });

      it('Should remove test from pending array', () =>
        new Promise(done => {
          runRunAsyncTest(testObject);
          setTimeout(
            context.bind({ stateRef }, () => {
              const [pendingState] = usePending();
              expect(pendingState.pending).not.toContain(testObject);
              done();
            })
          );
        }));

      describe('When test is canceled', () => {
        beforeEach(() => {
          testObject.cancel();
        });

        it.ctx('Should remove test from pending array', () => {
          const [pendingState] = usePending();
          expect(pendingState.pending).toEqual(
            expect.arrayContaining([testObject])
          );
          runRunAsyncTest(testObject);
          return new Promise(done => {
            setTimeout(() => {
              expect(pendingState).toEqual(
                expect.not.arrayContaining([testObject])
              );
              done();
            });
          });
        });
      });
    });

    describe('doneCallbacks', () => {
      let fieldCallback_1, fieldCallback_2, doneCallback;
      beforeEach(() => {
        fieldCallback_1 = jest.fn();
        fieldCallback_2 = jest.fn();
        doneCallback = jest.fn();

        context.run({ stateRef }, () => {
          const [, setTestCallbacks] = useTestCallbacks();

          setTestCallbacks(state => ({
            fieldCallbacks: {
              ...state.fieldCallbacks,
              [fieldName]: (state.fieldCallbacks[fieldName] || []).concat(
                fieldCallback_1,
                fieldCallback_2
              ),
            },
            doneCallbacks: state.doneCallbacks.concat(doneCallback),
          }));
        });
      });
      describe('When no remaining tests', () => {
        it('Should run all callbacks', () =>
          new Promise(done => {
            expect(fieldCallback_1).not.toHaveBeenCalled();
            expect(fieldCallback_2).not.toHaveBeenCalled();
            expect(doneCallback).not.toHaveBeenCalled();
            runRunAsyncTest(testObject);
            setTimeout(() => {
              expect(fieldCallback_1).toHaveBeenCalled();
              expect(fieldCallback_2).toHaveBeenCalled();
              expect(doneCallback).toHaveBeenCalled();
              done();
            });
          }));
      });

      describe('When there are more tests left', () => {
        beforeEach(() => {
          context.run({ stateRef }, () => {
            setPending(
              new VestTest({
                fieldName: 'pending_field',
                statement: STATEMENT,
                testFn: jest.fn(),
              })
            );
          });
        });

        it("Should only run current field's callbacks", () =>
          new Promise(done => {
            expect(fieldCallback_1).not.toHaveBeenCalled();
            expect(fieldCallback_2).not.toHaveBeenCalled();
            expect(doneCallback).not.toHaveBeenCalled();
            runRunAsyncTest(testObject);
            setTimeout(() => {
              expect(fieldCallback_1).toHaveBeenCalled();
              expect(fieldCallback_2).toHaveBeenCalled();
              expect(doneCallback).not.toHaveBeenCalled();
              done();
            });
          }));
      });

      describe('When test is canceled', () => {
        beforeEach(() => {
          testObject.cancel();
        });

        it('Should return without running any callback', () =>
          new Promise(done => {
            expect(fieldCallback_1).not.toHaveBeenCalled();
            expect(fieldCallback_2).not.toHaveBeenCalled();
            expect(doneCallback).not.toHaveBeenCalled();
            runRunAsyncTest(testObject);
            setTimeout(() => {
              expect(fieldCallback_1).not.toHaveBeenCalled();
              expect(fieldCallback_2).not.toHaveBeenCalled();
              expect(doneCallback).not.toHaveBeenCalled();
              done();
            });
          }));
      });
    });

    describe('testObject', () => {
      let testObjectCopy;

      beforeEach(() => {
        testObject.fail = jest.fn();
        testObjectCopy = _.cloneDeep(testObject);
      });

      if (testCase === CASE_PASSING) {
        it('Should keep test object unchanged', () =>
          new Promise(done => {
            runRunAsyncTest(testObject);
            setTimeout(() => {
              expect(testObject).toEqual(testObjectCopy);
              done();
            });
          }));

        it('Should return without calling testObject.fail', () =>
          new Promise(done => {
            runRunAsyncTest(testObject);
            setTimeout(() => {
              expect(testObject.fail).not.toHaveBeenCalled();
              done();
            });
          }));
      }

      if (testCase === CASE_FAILING) {
        it('Should call testObject.fail', () =>
          new Promise(done => {
            runRunAsyncTest(testObject);
            setTimeout(() => {
              expect(testObject.fail).toHaveBeenCalled();
              done();
            });
          }));

        describe('When rejecting with a message', () => {
          const rejectionString = 'rejection string';
          beforeEach(() => {
            testObject.asyncTest.catch(Function.prototype);
            testObject.asyncTest = Promise.reject(rejectionString);
          });

          it('Should set test statement to rejection string', () =>
            new Promise(done => {
              runRunAsyncTest(testObject);
              setTimeout(() => {
                expect(testObject.statement).toBe(rejectionString);
                done();
              });
            }));
        });
      }
    });
  }
);
