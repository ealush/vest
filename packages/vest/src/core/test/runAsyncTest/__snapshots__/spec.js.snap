// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`runAsyncTest: failing State updates Initial state matches snapshot (sanity) 1`] = `
Object {
  "doneCallbacks": Array [],
  "fieldCallbacks": Object {
    "field_1": Array [],
  },
  "groups": Object {},
  "lagging": Array [],
  "name": "suite_name_1",
  "pending": Array [
    VestTest {
      "asyncTest": Promise {},
      "failed": false,
      "fieldName": "field_1",
      "id": "11",
      "isWarning": false,
      "statement": "some statement string",
      "suiteId": "suiteId_1",
      "testFn": [Function],
    },
  ],
  "suiteId": "suiteId_1",
  "testObjects": Array [],
  "tests": Object {},
}
`;

exports[`runAsyncTest: passing State updates Initial state matches snapshot (sanity) 1`] = `
Object {
  "doneCallbacks": Array [],
  "fieldCallbacks": Object {
    "field_1": Array [],
  },
  "groups": Object {},
  "lagging": Array [],
  "name": "suite_name_1",
  "pending": Array [
    VestTest {
      "asyncTest": Promise {},
      "failed": false,
      "fieldName": "field_1",
      "id": "0",
      "isWarning": false,
      "statement": "some statement string",
      "suiteId": "suiteId_1",
      "testFn": [Function],
    },
  ],
  "suiteId": "suiteId_1",
  "testObjects": Array [],
  "tests": Object {},
}
`;
